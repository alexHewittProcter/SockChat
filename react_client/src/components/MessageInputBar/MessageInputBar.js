import React from 'react';
import propTypes from 'prop-types';
import InputGroup from 'react-bootstrap/InputGroup';
import Button from 'react-bootstrap/Button';
import FormControl from 'react-bootstrap/FormControl';

class MessageInputBar extends React.Component {
  constructor(props) {
    super(props);
    this.state = { value: '' };
    this.sendClick = this.sendClick.bind(this);
    this.inputKeyDown = this.inputKeyDown.bind(this);
    this.messageInput = React.createRef();
    this.focusInput = this.focusInput.bind(this);
  }
  render() {
    return (
      <div className="message-input-bar">
        <InputGroup>
          <FormControl
            placeholder="Message..."
            value={this.state.value}
            onChange={(event) => {
              this.setState({ value: event.target.value });
            }}
            onKeyDown={this.inputKeyDown}
            ref={this.messageInput}
            name="message"
          />
          <InputGroup.Append>
            <Button onClick={this.sendClick}>Send</Button>
          </InputGroup.Append>
        </InputGroup>
      </div>
    );
  }
  componentDidMount() {
    this.focusInput();
  }
  focusInput() {
    this.messageInput.current.focus();
  }
  inputKeyDown(event) {
    if (event.key === 'Enter') {
      this.sendClick();
    }
  }
  sendClick() {
    if (!!this.state.value) {
      this.props.messageSend(this.state.value);
      this.setState({ value: '' });
    }
  }
}

MessageInputBar.propTypes = { messageSend: propTypes.func };
MessageInputBar.defaultProps = {
  messageSend: (msg) => {
    console.log('Sent');
    console.log(msg);
  },
};

export default MessageInputBar;
