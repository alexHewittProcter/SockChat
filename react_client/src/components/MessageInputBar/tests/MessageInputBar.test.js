import React from 'react';
import { mount } from 'enzyme';
import MessageInputBar from '../MessageInputBar';

describe('MessageInputBar', () => {
  let messageInputBar;
  let messageSend;
  let props;

  let getComp = (selector) => messageInputBar.find(selector);
  let getInput = () => getComp('input[name="message"]');
  let getValue = (component) => component.prop('value');
  it('should render correctly', () => {
    messageInputBar = mount(<MessageInputBar />);
    expect(messageInputBar).toMatchSnapshot();
  });
  beforeEach(() => {
    messageSend = jest.fn();
    props = { messageSend };
    messageInputBar = mount(<MessageInputBar {...props} />);
  });
  describe('Message input', () => {
    it('should allow a user to input text/numbers', () => {
      expect(getValue(getInput())).toEqual('');
      getInput().simulate('change', { target: { name: 'message', value: 'Test input' } });
      expect(getValue(getInput())).toEqual('Test input');
    });
    describe('passing the inputted message when pressing the Enter key', () => {
      it('should not call the function when the input is empty', () => {
        expect(getValue(getInput())).toEqual('');
        getInput().simulate('keydown', { key: 'Enter' });
        expect(messageSend).toHaveBeenCalledTimes(0);
      });
      it('should call the function when the input is not empty', () => {
        getInput().simulate('change', { target: { name: 'message', value: 'Test input' } });
        getInput().simulate('keydown', { key: 'Enter' });
        expect(messageSend).toHaveBeenCalledTimes(1);
        expect(messageSend).toHaveBeenCalledWith('Test input');
      });
    });
  });
});
