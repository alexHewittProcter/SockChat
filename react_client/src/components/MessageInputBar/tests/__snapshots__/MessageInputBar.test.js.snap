// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageInputBar should render correctly 1`] = `
<div
  className="message-input-bar"
>
  <InputGroup>
    <div
      className="input-group"
    >
      <FormControl
        name="message"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Message..."
        value=""
      >
        <input
          className="form-control"
          name="message"
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Message..."
          value=""
        />
      </FormControl>
      <InputGroupAppend>
        <div
          className="input-group-append"
        >
          <Button
            active={false}
            disabled={false}
            onClick={[Function]}
            type="button"
            variant="primary"
          >
            <button
              className="btn btn-primary"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Send
            </button>
          </Button>
        </div>
      </InputGroupAppend>
    </div>
  </InputGroup>
</div>
`;
